Move CUDA-related stuff of TP agent to separate file

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59788 Prepare for TensorPipe separating its CUDA-specific headers
* **#59377 Move CUDA-related stuff of TP agent to separate file**
* #59376 Make CUDA serde support for TP agent pluggable

This PR demonstrates that now the CUDA parts of the TensorPipe agent just "plug on top" of the CPU-only parts. Thus ideally the CPU-only parts could go in libtorch while the CUDA-only parts could go in libtorch_cuda. Unfortunately we can't do that just yet, because the TensorPipe agent depends on c10d (for its Store and its ProcessGroup), which lives in libtorch_python.

Differential Revision: [D28796429](https://our.internmc.facebook.com/intern/diff/D28796429/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28796429/)!

lw committed:
Move CUDA-related stuff of TP agent to separate file

This PR demonstrates that now the CUDA parts of the TensorPipe agent just "plug on top" of the CPU-only parts. Thus ideally the CPU-only parts could go in libtorch while the CUDA-only parts could go in libtorch_cuda. Unfortunately we can't do that just yet, because the TensorPipe agent depends on c10d (for its Store and its ProcessGroup), which lives in libtorch_python.

Differential Revision: [D28796429](https://our.internmc.facebook.com/intern/diff/D28796429/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28796429/)!

[ghstack-poisoned]

lw committed:
Update on "Move CUDA-related stuff of TP agent to separate file"


This PR demonstrates that now the CUDA parts of the TensorPipe agent just "plug on top" of the CPU-only parts. Thus ideally the CPU-only parts could go in libtorch while the CUDA-only parts could go in libtorch_cuda. Unfortunately we can't do that just yet, because the TensorPipe agent depends on c10d (for its Store and its ProcessGroup), which lives in libtorch_python.

Differential Revision: [D28796429](https://our.internmc.facebook.com/intern/diff/D28796429/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28796429/)!

[ghstack-poisoned]

lw committed:
Update on "Move CUDA-related stuff of TP agent to separate file"


This PR demonstrates that now the CUDA parts of the TensorPipe agent just "plug on top" of the CPU-only parts. Thus ideally the CPU-only parts could go in libtorch while the CUDA-only parts could go in libtorch_cuda. Unfortunately we can't do that just yet, because the TensorPipe agent depends on c10d (for its Store and its ProcessGroup), which lives in libtorch_python.

Differential Revision: [D28796429](https://our.internmc.facebook.com/intern/diff/D28796429/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28796429/)!

[ghstack-poisoned]

lw committed:
Update on "Move CUDA-related stuff of TP agent to separate file"


This PR demonstrates that now the CUDA parts of the TensorPipe agent just "plug on top" of the CPU-only parts. Thus ideally the CPU-only parts could go in libtorch while the CUDA-only parts could go in libtorch_cuda. Unfortunately we can't do that just yet, because the TensorPipe agent depends on c10d (for its Store and its ProcessGroup), which lives in libtorch_python.

Differential Revision: [D28796429](https://our.internmc.facebook.com/intern/diff/D28796429/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28796429/)!

[ghstack-poisoned]

lw committed:
Update on "Move CUDA-related stuff of TP agent to separate file"


This PR demonstrates that now the CUDA parts of the TensorPipe agent just "plug on top" of the CPU-only parts. Thus ideally the CPU-only parts could go in libtorch while the CUDA-only parts could go in libtorch_cuda. Unfortunately we can't do that just yet, because the TensorPipe agent depends on c10d (for its Store and its ProcessGroup), which lives in libtorch_python.

Differential Revision: [D28796429](https://our.internmc.facebook.com/intern/diff/D28796429/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28796429/)!

[ghstack-poisoned]

cbalioglu reviewed (approved):
LGTM

Nit: you might also "fix" the loop at line 138 in `tensorpipe_agent.cpp` while you are here.

```c++
#include <c10/util/irange.h>

for(const auto idx : c10::irange(indexBitset.size())) {
  ...
}
```

lw committed:
Update on "Move CUDA-related stuff of TP agent to separate file"


This PR demonstrates that now the CUDA parts of the TensorPipe agent just "plug on top" of the CPU-only parts. Thus ideally the CPU-only parts could go in libtorch while the CUDA-only parts could go in libtorch_cuda. Unfortunately we can't do that just yet, because the TensorPipe agent depends on c10d (for its Store and its ProcessGroup), which lives in libtorch_python.

Differential Revision: [D28796429](https://our.internmc.facebook.com/intern/diff/D28796429/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28796429/)!

[ghstack-poisoned]

lw committed:
Update on "Move CUDA-related stuff of TP agent to separate file"


This PR demonstrates that now the CUDA parts of the TensorPipe agent just "plug on top" of the CPU-only parts. Thus ideally the CPU-only parts could go in libtorch while the CUDA-only parts could go in libtorch_cuda. Unfortunately we can't do that just yet, because the TensorPipe agent depends on c10d (for its Store and its ProcessGroup), which lives in libtorch_python.

Differential Revision: [D28796429](https://our.internmc.facebook.com/intern/diff/D28796429/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28796429/)!

[ghstack-poisoned]

lw committed:
Update on "Move CUDA-related stuff of TP agent to separate file"


This PR demonstrates that now the CUDA parts of the TensorPipe agent just "plug on top" of the CPU-only parts. Thus ideally the CPU-only parts could go in libtorch while the CUDA-only parts could go in libtorch_cuda. Unfortunately we can't do that just yet, because the TensorPipe agent depends on c10d (for its Store and its ProcessGroup), which lives in libtorch_python.

Differential Revision: [D28796429](https://our.internmc.facebook.com/intern/diff/D28796429/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28796429/)!

[ghstack-poisoned]

lw committed:
Update on "Move CUDA-related stuff of TP agent to separate file"


This PR demonstrates that now the CUDA parts of the TensorPipe agent just "plug on top" of the CPU-only parts. Thus ideally the CPU-only parts could go in libtorch while the CUDA-only parts could go in libtorch_cuda. Unfortunately we can't do that just yet, because the TensorPipe agent depends on c10d (for its Store and its ProcessGroup), which lives in libtorch_python.

Differential Revision: [D28796429](https://our.internmc.facebook.com/intern/diff/D28796429/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28796429/)!

[ghstack-poisoned]

lw committed:
Update on "Move CUDA-related stuff of TP agent to separate file"


This PR demonstrates that now the CUDA parts of the TensorPipe agent just "plug on top" of the CPU-only parts. Thus ideally the CPU-only parts could go in libtorch while the CUDA-only parts could go in libtorch_cuda. Unfortunately we can't do that just yet, because the TensorPipe agent depends on c10d (for its Store and its ProcessGroup), which lives in libtorch_python.

Differential Revision: [D28796429](https://our.internmc.facebook.com/intern/diff/D28796429/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28796429/)!

[ghstack-poisoned]

mrshenli reviewed (commented):
+1 to @cbalioglu's comment. Please consider `NOLINTNEXTLINE` if those are intentional.

lw committed:
Update on "Move CUDA-related stuff of TP agent to separate file"


This PR demonstrates that now the CUDA parts of the TensorPipe agent just "plug on top" of the CPU-only parts. Thus ideally the CPU-only parts could go in libtorch while the CUDA-only parts could go in libtorch_cuda. Unfortunately we can't do that just yet, because the TensorPipe agent depends on c10d (for its Store and its ProcessGroup), which lives in libtorch_python.

Differential Revision: [D28796429](https://our.internmc.facebook.com/intern/diff/D28796429/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28796429/)!

[ghstack-poisoned]
